### PR TITLE
feat: add logic to build an execution payload when proposing blocks.

### PR DESCRIPTION
**Pull request by mpaulucci**
**Motivation**
Being able to build an execution block

**Description**
Naive implementation of calling the execution layer to get an execution payload

### CODE REVIEW

File: lib/lambda_ethereum_consensus/validator/validator.ex [hunk 1]
```diff
@@ -408,35 +411,69 @@ defmodule LambdaEthereumConsensus.Validator do
...
-    {:ok, ssz_encoded} = Ssz.to_ssz(signed_block)
-    {:ok, encoded_msg} = :snappyer.compress(ssz_encoded)
-    fork_context = BeaconChain.get_fork_digest() |> Base.encode16(case: :lower)
+    with {:ok, payload_id} <-
+           ExecutionClient.notify_forkchoice_updated(
+             forkchoice_state,
+             payload_attributes
+           ),
+         # TODO: find a better way
+         :ok <- Process.sleep(3000),
```
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/validator/validator.ex:442]: We should retry this a few times:

```elixir
Stream.repeatedly(fn -> ExecutionClient. get_payload(payload_id) end)
|> Stream.zip(Stream.timer(timeout_in_ms) |> Stream.cycle())
|> Stream.map(fn {v, 0} -> v end) 
|> Stream.take(number_of_retries)
|> Enum.find(&match?({:ok, _}, &1})
```

File: lib/lambda_ethereum_consensus/validator/validator.ex [hunk 2]
```diff
@@ -408,35 +411,69 @@ defmodule LambdaEthereumConsensus.Validator do
...
 
-    # TODO: we might want to send the block to ForkChoice
+  defp propose(%{root: head_root, validator: %{index: index, privkey: privkey}}, proposed_slot) do
+    head_state = BlockStates.get_state!(head_root)
 
-    case Libp2pPort.publish("/eth2/#{fork_context}/beacon_block/ssz_snappy", encoded_msg) do
-      :ok -> Logger.info("[Validator] Proposed block for slot #{proposed_slot}")
-      _ -> Logger.error("[Validator] Failed to publish block for slot #{proposed_slot}")
+    with {:ok, execution_payload} <-
+           build_execution_block(head_state |> process_slots(proposed_slot), head_root) do
```
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/validator/validator.ex:453]: nit
```suggestion
           head_state |> process_slots(proposed_slot) |> build_execution_block(head_root) do
```

File: lib/lambda_ethereum_consensus/execution/execution_client.ex [hunk 3]
```diff
@@ -11,6 +12,18 @@ defmodule LambdaEthereumConsensus.Execution.ExecutionClient do
...
   @type execution_status :: :optimistic | :valid | :invalid | :unknown
 
+  @spec get_payload(any()) :: {:error, any()} | {:ok, any()}
+  def get_payload(payload_id) do
+    case EngineApi.get_payload(payload_id) do
+      {:ok, %{"execution_payload" => raw_payload}} ->
+        {:ok, parse_raw_payload(raw_payload)}
+
+      {:error, reason} ->
+        Logger.warning("Error when calling get payload: #{inspect(reason)}")
```
Comment (MegaRedHand) [inline lib/lambda_ethereum_consensus/execution/execution_client.ex:22]: This should be "error" level

**Review by MegaRedHand — APPROVED** [review]
Left some nits. LGTM otherwise

File: lib/lambda_ethereum_consensus/validator/validator.ex (hunk 1 above)
Comment (MegaRedHand) [reply lib/lambda_ethereum_consensus/validator/validator.ex:442]: But isn't that time variable? What happens if we call this without enough sleep?

Comment (mpaulucci) [reply lib/lambda_ethereum_consensus/validator/validator.ex:442]: There will be fewer/no transactions but it will succeed

Comment (mpaulucci) [reply lib/lambda_ethereum_consensus/validator/validator.ex:442]: When we make this code more sophisticated, we need to balance a time that should be long enough to let the execution pack as many transactions as possible (more fees for us) while giving enough time to propagate the block and have it included

Comment (mpaulucci) [reply lib/lambda_ethereum_consensus/validator/validator.ex]: Added this comment in code.

**Pull request merged**